velox/functions/iceberg/BucketFunction.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/iceberg/BucketFunction.h"
+#include "velox/functions/Macros.h"
+#include "velox/functions/Registerer.h"
+#include "velox/functions/iceberg/util/Murmur3_32HashFunction.h"
+#include "velox/type/Timestamp.h"
+
+namespace facebook::velox::functions::iceberg {
+namespace {
+
+FOLLY_ALWAYS_INLINE int apply(int numBuckets, int hashedValue) {
+  return (hashedValue & INT_MAX) % numBuckets;
+}
+
+template <typename T>
+struct BucketDecimalFunction {
```

Comment:
I notice the bucket functions are put in a common folder, and does it imply that the bucket functions are independent from the execution engine? Would there be any difference between Presto and Spark?

Reply:
Presto engine does not support bucket function now but may support in the future, trino has supported them. Trino implementation https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTransforms.java, It uses io.airlift.slice.Murmur3Hash32, not sure why it does not use guava hash as spark in Iceberg Repo. But I suppose the implementation is same, otherwise spark cannot read the data written by trino.

Reply:
I'm thinking if separate folder is needed for presto and spark bucket functions for both the doc and the implementation. Current implementation does not belong to the common folder as well.

Reply:
I double check the trino implemation, it is same with Spark, so this implementation can be used for both trino(presto), spark, https://github.com/airlift/slice/blob/master/src/main/java/io/airlift/slice/Murmur3Hash32.java